### PR TITLE
replace Bluebird.timeout

### PR DESCRIPTION
**Pull request by maximelkin**
Breaking changes: TImeoutError now not extends Bluebird.TimeoutError
also this error exported explicitly

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,6 +39,7 @@
     "liftoff": "3.1.0",
     "lodash": "^4.17.15",
     "mkdirp": "^0.5.1",
+    "p-timeout": "^3.2.0",
```
Comment (kibertoad) [inline package.json]: Considering how small p-yimeout is, maybe it would make sense to inline implementation and use native finally instead of p-finally?

Comment (kibertoad) [reply package.json]: no strong opinion on that, though, sindresorhus libs are legit.

**Comment by maximelkin** [comment]
Does we still support node 8?

File: package.json (hunk 1 above)
Comment (maximelkin) [reply package.json]: Done

**Comment by kibertoad** [comment]
@maximelkin Yeah, ideally we want to support Node 8 for a bit longer. If that makes implementation complex, we can stick with p-timeout, it's not a big deal.

**Comment by maximelkin** [comment]
This can make harder to move from bluebird completely, because Promise.finally only from 10 node

**Comment by kibertoad** [comment]
@maximelkin doesn't p-timeout rely on p-finally which supports node 8?

**Comment by maximelkin** [comment]
Yes, it is
But p-timeout dependency removed

**Comment by kibertoad** [comment]
@maximelkin I'd much rather use p-finally directly or p-timeout rather than keep bluebird or break node 8  compatibility.

**Comment by maximelkin** [comment]
p-timeout required finally because of support  .cancel, we dont need it, so no troubles here

**Comment by kibertoad** [comment]
ah, great :)

**Comment by kibertoad** [comment]
@maximelkin Since https://github.com/knex/knex/pull/3639 is dropping support for Node 8 anyway, I guess this PR can do the same and we can merge both during next breaking release.

**Comment by maximelkin** [comment]
Not directly related question:
bluebird will be replaced completely to es6 promises, or user may choose api/internal promises?

**Comment by elhigu** [comment]
> bluebird will be replaced completely to es6 promises, or user may choose api/internal promises?

Knex will use only native promises and will not allow to choose the implementation.

**Comment by maximelkin** [comment]
@kibertoad This pr actually not breaking node 8 compatibility, maybe we can merge it?

**Comment by kibertoad** [comment]
@maximelkin There's a new unhandled Rejection on Node 8 happening ever since this was merged:

```
{ reason: Error: Transaction rejected with non-error: undefined
    at q.trxClient.query.catch.then.res (/home/travis/build/knex/knex/lib/transaction.js:15:2936)
From previous event:
    at Transaction._promise.acquireConnection.connection (/home/travis/build/knex/knex/lib/transaction.js:8:1205)
    at Bluebird.then.then (/home/travis/build/knex/knex/lib/transaction.js:18:1002)
From previous event:
    at Transaction.acquireConnection (/home/travis/build/knex/knex/lib/transaction.js:18:886)
    at new Transaction (/home/travis/build/knex/knex/lib/transaction.js:8:834)
    at Client_SQLite3.transaction (/home/travis/build/knex/knex/lib/client.js:7:2337)
    at Function.transaction (/home/travis/build/knex/knex/lib/util/make-knex.js:6:103)
    at Context.it (/home/travis/build/knex/knex/test/unit/knex.js:403:28)
    at callFn (/home/travis/build/knex/knex/node_modules/mocha/lib/runnable.js:395:21)
    at Test.Runnable.run (/home/travis/build/knex/knex/node_modules/mocha/lib/runnable.js:382:7)
    at Runner.runTest (/home/travis/build/knex/knex/node_modules/mocha/lib/runner.js:541:10)
    at /home/travis/build/knex/knex/node_modules/mocha/lib/runner.js:667:12
    at next (/home/travis/build/knex/knex/node_modules/mocha/lib/runner.js:450:14)
    at /home/travis/build/knex/knex/node_modules/mocha/lib/runner.js:460:7
    at next (/home/travis/build/knex/knex/node_modules/mocha/lib/runner.js:362:14)
    at Immediate._onImmediate (/home/travis/build/knex/knex/node_modules/mocha/lib/runner.js:428:5)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5) }
```

Wonder if there's a gap in our timeout implementation and if sindresorhus one would be more reliable.

**Comment by maximelkin** [comment]
@kibertoad 
I think it somehow connected with refactoring from bluebird pattern-match by class .catch to es6 .catch

**Pull request merged**